lib/src/user.dart

```diff
@@ -27,12 +32,96 @@ import 'native/realm_core.dart';
 class User {
   final UserHandle _handle;
 
-  User._(this._handle);
+  /// The [App] with which the user is associated with.
+  final App app;
+
+  User._(this.app, this._handle);
+
+  /// The custom user data associated with this user.
+  dynamic get customData {
+    final data = realmCore.userGetCustomData(this);
+    return jsonDecode(data);
+  }
+
+  /// Refreshes the custom data for a this [User].
+  Future<dynamic> refreshCustomData() async {
+    await realmCore.userRefreshCustomData(app, this);
+    return customData;
+  }
+
+  /// Links this [User] with a new [User] identity represented by the given credentials.
+  ///
+  /// Linking a user with more credentials, mean the user can login either of these credentials. It also makes it possible to "upgrade" an anonymous user
+  /// by linking it with e.g. Email/Password credentials.
+  /// Note: It is not possible to link two existing users of MongoDB Realm. The provided credentials must not have been used by another user.
+  Future<User> linkCredentials(Credentials credentials) async {
+    final userHandle = await realmCore.userLinkCredentials(app, this, credentials);
+    return UserInternal.create(app, userHandle);
+  }
+
+  /// The current state of this [User].
+  UserState get state {
+    final nativeState = realmCore.userGetState(this);
```

Comment:
Should we change `realmCore.userGetState` to return `UserState` rather than an int?

Reply:
I moved the conversion to realm_core